### ISSUE TITLE
builds in freshly created projects take a long time to complete, waiting on signing

### ISSUE THREAD

**Issue by evgeni** [fedora-copr/copr#3050]
Example build: https://copr.fedorainfracloud.org/coprs/g/theforeman/plugins-nightly-staging-scratch-994081e7-0d39-5cac-9bb4-b71f53759035/build/6726503/

In @theforeman we create projects on demand to do "scratch" builds (so these do not pollute the main repo).

In the past, this had the problem that if you create a project and schedule a build very quickly afterwards, you run into a state where the gpg key is not yet created, and signing fails. This was fixed in https://github.com/fedora-copr/copr/pull/2954 by adding a wait/retry thing (forgive me my technical ignorance here, I do not know Copr very well).

However, now we have observed builds sometimes taking *a long* time, like hours.

Looking [at the logs](https://download.copr.fedorainfracloud.org/results/@theforeman/plugins-nightly-staging-scratch-994081e7-0d39-5cac-9bb4-b71f53759035/rhel-8-x86_64/06726503-rubygem-foreman_bootdisk/backend.log.gz), one sees a huge gap while it waits on being able to sign things:
```
[2023-12-05 13:27:51,746][  INFO][PID:1586497] Going to sign pkgs from source: 6726503-rhel-8-x86_64 in chroot: /var/lib/copr/public_html/results/@theforeman/plugins-nightly-staging-scratch-994081e7-0d39-5cac-9bb4-b71f53759035/rhel-8-x86_64
[2023-12-05 13:27:51,747][  INFO][PID:1586497] Calling '/bin/sign -u @theforeman#plugins-nightly-staging-scratch-994081e7-0d39-5cac-9bb4-b71f53759035@copr.fedorahosted.org -p' (attempt #1)
[2023-12-05 13:27:52,315][WARNING][PID:1586497] Command '/bin/sign -u @theforeman#plugins-nightly-staging-scratch-994081e7-0d39-5cac-9bb4-b71f53759035@copr.fedorahosted.org -p' failed with: unknown key: @theforeman#plugins-nightly-staging-scratch-994081e7-0d39-5cac-9bb4-b71f53759035@copr.fedorahosted.org
[2023-12-05 13:27:52,316][WARNING][PID:1586497] Going to sleep 20s and re-try.
[2023-12-05 13:28:12,317][  INFO][PID:1586497] Calling '/bin/sign -u @theforeman#plugins-nightly-staging-scratch-994081e7-0d39-5cac-9bb4-b71f53759035@copr.fedorahosted.org -p' (attempt #2)
[2023-12-05 13:28:13,074][WARNING][PID:1586497] Command '/bin/sign -u @theforeman#plugins-nightly-staging-scratch-994081e7-0d39-5cac-9bb4-b71f53759035@copr.fedorahosted.org -p' failed with: unknown key: @theforeman#plugins-nightly-staging-scratch-994081e7-0d39-5cac-9bb4-b71f53759035@copr.fedorahosted.org
[2023-12-05 13:28:13,075][WARNING][PID:1586497] Going to sleep 20s and re-try.
[2023-12-05 13:28:33,076][  INFO][PID:1586497] Calling '/bin/sign -u @theforeman#plugins-nightly-staging-scratch-994081e7-0d39-5cac-9bb4-b71f53759035@copr.fedorahosted.org -p' (attempt #3)
[2023-12-05 13:28:33,941][WARNING][PID:1586497] Command '/bin/sign -u @theforeman#plugins-nightly-staging-scratch-994081e7-0d39-5cac-9bb4-b71f53759035@copr.fedorahosted.org -p' failed with: unknown key: @theforeman#plugins-nightly-staging-scratch-994081e7-0d39-5cac-9bb4-b71f53759035@copr.fedorahosted.org
[2023-12-05 13:28:33,942][WARNING][PID:1586497] Going to sleep 20s and re-try.
[2023-12-06 14:18:57,914][  INFO][PID:1586497] Calling '/bin/sign -4 -h sha256 -u @theforeman#plugins-nightly-staging-scratch-994081e7-0d39-5cac-9bb4-b71f53759035@copr.fedorahosted.org -r /var/lib/copr/public_html/results/@theforeman/plugins-nightly-staging-scratch-994081e7-0d39-5cac-9bb4-b71f53759035/rhel-8-x86_64/06726503-rubygem-foreman_bootdisk/rubygem-foreman_bootdisk-doc-21.2.1-1.fm3_10.el8.noarch.rpm' (attempt #1)
[2023-12-06 14:18:58,713][  INFO][PID:1586497] signed rpm: /var/lib/copr/public_html/results/@theforeman/plugins-nightly-staging-scratch-994081e7-0d39-5cac-9bb4-b71f53759035/rhel-8-x86_64/06726503-rubygem-foreman_bootdisk/rubygem-foreman_bootdisk-doc-21.2.1-1.fm3_10.el8.noarch.rpm
[2023-12-06 14:18:58,714][  INFO][PID:1586497] Calling '/bin/sign -4 -h sha256 -u @theforeman#plugins-nightly-staging-scratch-994081e7-0d39-5cac-9bb4-b71f53759035@copr.fedorahosted.org -r /var/lib/copr/public_html/results/@theforeman/plugins-nightly-staging-scratch-994081e7-0d39-5cac-9bb4-b71f53759035/rhel-8-x86_64/06726503-rubygem-foreman_bootdisk/rubygem-foreman_bootdisk-21.2.1-1.fm3_10.el8.noarch.rpm' (attempt #1)
[2023-12-06 14:18:59,541][  INFO][PID:1586497] signed rpm: /var/lib/copr/public_html/results/@theforeman/plugins-nightly-staging-scratch-994081e7-0d39-5cac-9bb4-b71f53759035/rhel-8-x86_64/06726503-rubygem-foreman_bootdisk/rubygem-foreman_bootdisk-21.2.1-1.fm3_10.el8.noarch.rpm
[2023-12-06 14:18:59,543][  INFO][PID:1586497] Calling '/bin/sign -4 -h sha256 -u @theforeman#plugins-nightly-staging-scratch-994081e7-0d39-5cac-9bb4-b71f53759035@copr.fedorahosted.org -r /var/lib/copr/public_html/results/@theforeman/plugins-nightly-staging-scratch-994081e7-0d39-5cac-9bb4-b71f53759035/rhel-8-x86_64/06726503-rubygem-foreman_bootdisk/rubygem-foreman_bootdisk-21.2.1-1.fm3_10.el8.src.rpm' (attempt #1)
[2023-12-06 14:19:00,328][  INFO][PID:1586497] signed rpm: /var/lib/copr/public_html/results/@theforeman/plugins-nightly-staging-scratch-994081e7-0d39-5cac-9bb4-b71f53759035/rhel-8-x86_64/06726503-rubygem-foreman_bootdisk/rubygem-foreman_bootdisk-21.2.1-1.fm3_10.el8.src.rpm
[2023-12-06 14:19:00,329][  INFO][PID:1586497] Sign done
```

**Comment by evgeni** [comment]
I've seen this behavior every now and then in the last ~week (and probably contributed to the overall load by running many builds in Foreman).

I'll monitor this on our side a bit and report back if things stabilized or not.

Thanks for running copr!

**Comment by evgeni** [comment]
Did y'all ever consider allowing users to disable signing? I think for our (crude) usecase it would be totally OK to disable signing all together because the results are rarely really used and if they are only for a quick test on a non prod system.

**Comment by praiskup** [comment]
So far, AFAIR, all the intentions were about making the signing process a non-problem, rather than allowing users to opt-out.  Normally "we are there" and signing doesn't hurt.  The fire we had to put down yesterday was rather extreme.

If you think it is worth having the option though, RFE ticket is welcome (patch even more).

**Comment by evgeni** [comment]
I like non-problems ;)
But also, the way we use copr currently puts a lot of load on your signing infra without any benefit as the scratch repos are never used for anything but "will it build". And wasting resources seems odd :)

I'll open a separate RFE for that.

**Comment by nikromen** [comment]
revert hotfix in mini keygen release

**Comment by praiskup** [comment]
I think this is resolved in production, but we should check that the hot-fix is in place (or possibly do the mini-release).

**Comment by praiskup** [comment]
Ok, in the worst case this will be resolved in #3151
